source/routing/next_stop_time.cpp

```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
     return first_discrete_st_pair;
 }
 
+
+template<typename F>
+static void vj_loop(const nt::DiscreteVehicleJourney*, F f) {
+    // Discrete Vehicle Journey doesn't need to loop
+    f(0);
+}
+
+template<typename F>
+static void vj_loop(const nt::FrequencyVehicleJourney* vj, F f) {
+    for (auto freq_shift = vj->start_time; freq_shift <= vj->end_time; freq_shift+= vj->headway_secs) {
+        f(freq_shift);
+    }
+}
+
+using DtSt = std::pair<DateTime, const type::StopTime*>;
+template<typename VJ_T>
+static void fill_cache(const DateTime from,
+        const DateTime to,
+        const type::RTLevel rt_level,
+        const type::AccessibiliteParams& accessibilite_params,
+        const JourneyPattern& jp,
+        const std::vector<const VJ_T*>& vjs,
+        IdxMap<JourneyPatternPoint, std::vector<DtSt>>& arrival,
+        IdxMap<JourneyPatternPoint, std::vector<DtSt>>& departure){
+
+    const int to_int = static_cast<int>(DateTimeUtils::date(to));
+    int from_int = static_cast<int>(DateTimeUtils::date(from)) - 1;
+    from_int = from_int > 0 ? from_int : 0;
+
+    for (const auto* vj :  vjs) {
+        if (! vj->accessible(accessibilite_params.vehicle_properties)) {
+            continue;
+        }
+        // test validity pattern of vj
+        const auto* vp = vj->validity_patterns[rt_level];
+        for (int day = from_int; day <=  to_int ; ++day) {
+            if (! vp->check(day)) {
+                continue;
+            }
+            const auto shift = navitia::DateTimeUtils::SECONDS_PER_DAY * day;
+            size_t i= 0;
+            for (const auto& st : vj->stop_time_list) {
+                auto jpp_idx = jp.jpps[i];
+                auto loop_impl = [&](DateTime freq_shift){
+                    if (st.drop_off_allowed()) {
+                        auto arrival_time = st.arrival_time + shift + freq_shift;
+                        if (from <= arrival_time && arrival_time <= to) {
+                            arrival[jpp_idx].emplace_back(arrival_time,&st);
+                        }
+                    }
+                    if (st.pick_up_allowed()) {
+                        auto departure_time = st.departure_time + shift + freq_shift;
+                        if (departure_time <= to && from <= departure_time ) {
+                            departure[jpp_idx].emplace_back(departure_time,&st);
+                        }
+                    }
+                };
+                // loop is done differently according to the type of Vehicle Journey
+                vj_loop(vj, loop_impl);
+                ++i;
+            }
+        }
+    }
+}
+
+void CachedNextStopTime::load(const type::Data& data,
+        const DateTime from,
+        const DateTime to,
+        const type::RTLevel rt_level,
+        const type::AccessibiliteParams& accessibilite_params) {
+    const auto& jp_container = data.dataRaptor->jp_container;
+
+    departure.assign(jp_container.get_jpps_values());
```

Comment:
Those maps will be sparse... there might be a perf improvement in avoiding this, but I'm really not sure.
As this is critical, a bench on that may be a good idea. Any opinion @TeXitoi ?
We have to check what we do when no departure exists.